### ISSUE TITLE
twistd imports mktap, causing a PendingDeprecationWarning

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3127]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3127|
|Type|defect|
|Created|2008-03-26 16:01:52Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-non-deprecation-3127-2|

It's importing the mktap module for a small utility function called "getid". This relationship could be easily inverted, to avoid the warning.

<details><summary>Searchable metadata</summary>

```
trac-id__3127 3127
type__defect defect
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_twistd_non_deprecation_3127_2 branches/twistd-non-deprecation-3127-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1206547312000000 1206547312000000
changetime__1207670006000000 1207670006000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23139]) Branching to 'twistd-non-deprecation-3127'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I put `getid` in `twisted.scripts.twistd` so that `mktap` wouldn't have to import `_twistd_unix` which seemed like it might be a bad idea.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

 * I don't think that `twisted.scripts.twistd` is the right place for this. To be clear, I think getid can be more generally useful, so it should somewhere in twisted.python. For now I don't see something better than `twisted.python.util`, but you may think of something else.

 * There are no direct tests of getid.

 * Maybe the warning could be changed to a `DeprecationWarning` if everything works fine.

 * a legacy alias `getid = _getid` should be created in mktap

 * I don't really understand why you made all these changes to _twistd_unix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [23157]) move functions which don't really need to be methods for the tests back to free functions

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [23158]) add a test for twisted.scripts.mktap.getid

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Addressed all the points, I think.

As for the changes to `_twistd_unix`, I reverted some of them but several free-functions remain methods of the `ApplicationRunner` subclass.  I did this so that the unit tests could override this behavior in a subclass rather than having to patch the module object.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

 * in util.py, the docstring of initgroups could be reformated
 * I don't think the copyright notice of test_util.py should contain all the years. 2001-2008 is fine enough
 * I'm not sure that moving the tests in scripts/test is a good idea. First, app.py mentions the old location. Then, lots of code in these tests actually test app.py, and not twistd. Maybe we want to split these tests in two.
 * in test_twistd, the skip on setuid is duplicated

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [23239]) docstring fixes

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23241]) Format copyright the same way as the rest of the codebase

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [23242]) Remove setuid skip duplication

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [23243]) Branching to 'twistd-non-deprecation-3127-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [23245]) reality

refs [#3127](https://github.com/twisted/twisted/issues/3127)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner|
|-|-|

All fixed, I think.

Splitting up `test_twistd.py` is probably a good idea, but I just put the module back where it was for now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

twisted/scripts/test/test_twistd.py is still there. Except that, that looks good, so please merge once that file removed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Got rid of the extra `test_twistd.py`, changed one last reference to it in `test_mktap.py` to refer to the `twisted.test.test_twistd` module instead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [23252]) Merge twistd-non-deprecation-3127-2

Author: exarkun
Reviewer: therve
Fixes [#3127](https://github.com/twisted/twisted/issues/3127)

Promote the PendingDeprecationWarning in twisted.scripts.mktap to a
DeprecationWarning.  Change twisted.scripts._twistd_unix so that it
does not import twisted.scripts.mktap (triggering the deprecation),
and change the tap plugin definitions to use a new, public API for
creating IServiceMaker plugins instead of using the private API from
twisted.scripts.mktap.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
